### PR TITLE
[JBPM-8936] ConcurrentModificationException when retrieving server template

### PR DESCRIPTION
**Pull request by elguardian**
different instances of the server template through the storage.

### CODE REVIEW

**Comment by elguardian** [comment]
this is ok to test.

**Comment by elguardian** [comment]
Jenkins retest this please.

**Comment by sutaakar** [comment]
Jenkins retest this

File: kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ContainerSpec.java [hunk 1]
```diff
@@ -44,6 +46,13 @@ public ReleaseId getReleasedId() {
     public ContainerSpec() {
     }
 
+    public ContainerSpec(ContainerSpec other) {
+        super(other.getId(), other.getContainerName(), other.getServerTemplateKey());
+        this.releasedId = other.releasedId;
+        this.status = other.status;
+        this.configs = other.getConfigs().isEmpty() ? Collections.emptyMap() : new EnumMap<>(other.getConfigs());
```
Comment (MarianMacik) [inline kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ContainerSpec.java]: Please change the line 38 to also use `EnumMap<>` for consistency.

File: kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/storage/InMemoryKieServerTemplateStorage.java [hunk 2]
```diff
@@ -54,17 +58,44 @@ public ServerTemplate store(ServerTemplate serverTemplate) {
 
     @Override
     public List<ServerTemplate> load() {
-        return new ArrayList<ServerTemplate>(store.values());
+        synchronized (store) {
```
Comment (MarianMacik) [inline kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/storage/InMemoryKieServerTemplateStorage.java]: I guess we no longer need `store` to be of `ConcurrentHashMap` implementation if we are fully locking it.

**Comment by elguardian** [comment]
Jenkins retest this please.

File: kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ContainerSpec.java (hunk 1 above)
Comment (MarianMacik) [reply kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ContainerSpec.java]: Still uses HashMap.

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by elguardian** [comment]
jenkins retest this

**Comment by elguardian** [comment]
@MarianMacik sonar cloud issues are not relevant. Do you have anything else to add ?

File: kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/storage/InMemoryKieServerTemplateStorage.java [hunk 3]
```diff
@@ -54,17 +60,44 @@ public ServerTemplate store(ServerTemplate serverTemplate) {
...
+        if (current == null) {
+            return null;
+        }
+
+        List<ContainerSpec> specs = current.getContainersSpec().stream().map(ContainerSpec::new).collect(Collectors.toList());
+
+        ServerTemplate serverTemplate = new ServerTemplate(current.getId(),
+                                                           current.getName(),
+                                                           new ArrayList<>(current.getCapabilities()),
+                                                           current.getConfigs().isEmpty() ? new EnumMap<>(Capability.class) : new EnumMap<>(current.getConfigs()),
```
Comment (MarianMacik) [inline kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/storage/InMemoryKieServerTemplateStorage.java]: Here we store the `EnumMap` instance but in the original state of an object there is an instance of `HashMap` created. https://github.com/kiegroup/droolsjbpm-integration/blob/1b42ad1f1009442f13a500842c97b16e19d06648/kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ServerTemplate.java#L40

File: kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ContainerSpec.java [hunk 4]
```diff
@@ -44,6 +45,13 @@ public ReleaseId getReleasedId() {
     public ContainerSpec() {
     }
 
+    public ContainerSpec(ContainerSpec other) {
+        super(other.getId(), other.getContainerName(), other.getServerTemplateKey());
+        this.releasedId = other.releasedId;
+        this.status = other.status;
+        this.configs = other.getConfigs().isEmpty() ? new EnumMap<Capability, ContainerConfig>(Capability.class) : new EnumMap<>(other.getConfigs());
```
Comment (MarianMacik) [inline kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ContainerSpec.java]: Here we use `EnumMap` but the initializer is set to `HashMap`, see:
https://github.com/kiegroup/droolsjbpm-integration/blob/1b42ad1f1009442f13a500842c97b16e19d06648/kie-server-parent/kie-server-controller/kie-server-controller-api/src/main/java/org/kie/server/controller/api/model/spec/ContainerSpec.java#L36

**Comment by elguardian** [comment]
Jenkins execute full downstream build

**Comment by elguardian** [comment]
Jenkins retest this please.

**Comment by elguardian** [comment]
Jenkins retest this please.

**Comment by elguardian** [comment]
Jenkins retest this please.

**Pull request merged**